daemon/k8s_watcher.go

```diff
@@ -93,50 +93,59 @@ var (
 
 	k8sCM = controller.NewManager()
 
-	ruleRevCache ruleRevisionCache
+	importMetadataCache ruleImportMetadataCache
 )
 
-// ruleRevisionCache maps the unique identifier of a CiliumNetworkPolicy
-// (namespace and name) to the policy revision number of the agent's policy
-// repository at the time said rule was imported.
-type ruleRevisionCache struct {
-	mutex               lock.RWMutex
-	ruleRevisionMapping map[string]uint64
+// ruleImportMetadataCache maps the unique identifier of a CiliumNetworkPolicy
+// (namespace and name) to metadata about the importing of the rule into the
+// agent's policy repository at the time said rule was imported (revision
+// number, and if any error occurred while importing).
+type ruleImportMetadataCache struct {
+	mutex                 lock.RWMutex
+	ruleImportMetadataMap map[string]policyImportMetadata
 }
 
-func initRuleRevCache() {
-	ruleRevCache.ruleRevisionMapping = make(map[string]uint64)
+type policyImportMetadata struct {
+	revision          uint64
+	policyImportError error
 }
 
-func (r *ruleRevisionCache) upsert(cnp *cilium_v2.CiliumNetworkPolicy, revision uint64) {
+func initRuleMetadataCache() {
+	importMetadataCache.ruleImportMetadataMap = make(map[string]policyImportMetadata)
+}
+
+func (r *ruleImportMetadataCache) upsert(cnp *cilium_v2.CiliumNetworkPolicy, revision uint64, importErr error) {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
 	if cnp == nil {
 		return
 	}
 	podNSName := k8sUtils.GetObjNamespaceName(&cnp.ObjectMeta)
-	r.ruleRevisionMapping[podNSName] = revision
+	r.ruleImportMetadataMap[podNSName] = policyImportMetadata{
```

Comment:
Did you consider storing the CNPNS itself instead of creating a new structure to hold this data?

Reply:
I didn't want to do that for multiple reasons - one of which is because this only caches data that is assured to be true *and not change* for a given `CiliumNetworkPolicy` on this node apart from addition / deletion of a rule (i.e., not within the controller for managing the status of the rule). The revision at time of importing the policy as well as whether there was an error importing the given rule will never change in a way that will cause race conditions, because importing the policy / creation of the resultant policy import error are performed synchronously outside of the controller to sync the `CiliumNetworkPolicyNodeStatus`. See my explanation here 
https://github.com/cilium/cilium/pull/5192#discussion_r210093293 for more context around why I did it this way. I did not want to add logic for updating this from the controller (which updates the node status) because the controller runs asynchronously, and calls to update the annotations may occur while the controller is running, which may result in an out-of-date `CiliumNetworkPolicyNodeStatus` being cached locally. 

TL;DR: I viewed this as the safest way of fixing this issue in the context of the 1.2 release due to the fact that these two attributes (revision, import error) are created outside of the controller in a synchronous matter, as opposed to the `CiliumNetworkPolicyNodeStatus` itself.